x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextInferenceResultFieldMapper.java

```diff
@@ -239,29 +242,51 @@ private static void parseFieldInferenceResultElement(
             }
         }
 
-        Set<String> requiredSubfields = REQUIRED_SUBFIELDS_MAP.get(subfieldPath);
-        if (requiredSubfields != null && visitedSubfields.containsAll(requiredSubfields) == false) {
-            Set<String> missingSubfields = requiredSubfields.stream()
+        if (visitedSubfields.containsAll(REQUIRED_SUBFIELDS) == false) {
+            Set<String> missingSubfields = REQUIRED_SUBFIELDS.stream()
                 .filter(s -> visitedSubfields.contains(s) == false)
                 .collect(Collectors.toSet());
             throw new DocumentParsingException(parser.getTokenLocation(), "Missing required subfields: " + missingSubfields);
         }
     }
 
-    private static NestedObjectMapper createNestedObjectMapper(
+    private static NestedObjectMapper createInferenceResultsObjectMapper(
         DocumentParserContext context,
         MapperBuilderContext mapperBuilderContext,
-        String fieldName
+        String fieldName,
+        ModelSettings modelSettings
     ) {
         IndexVersion indexVersionCreated = context.indexSettings().getIndexVersionCreated();
-        ObjectMapper.Builder sparseVectorMapperBuilder = new ObjectMapper.Builder(
-            SPARSE_VECTOR_SUBFIELD_NAME,
-            ObjectMapper.Defaults.SUBOBJECTS
-        ).add(
-            new BooleanFieldMapper.Builder(SparseEmbeddingResults.Embedding.IS_TRUNCATED, ScriptCompiler.NONE, false, indexVersionCreated)
-        ).add(new SparseVectorFieldMapper.Builder(SparseEmbeddingResults.Embedding.EMBEDDING));
+        FieldMapper.Builder resultsBuilder;
+        if (modelSettings.taskType() == TaskType.SPARSE_EMBEDDING) {
+            resultsBuilder = new SparseVectorFieldMapper.Builder(INFERENCE_CHUNKS_RESULTS);
+        } else if (modelSettings.taskType() == TaskType.TEXT_EMBEDDING) {
+            DenseVectorFieldMapper.Builder denseVectorMapperBuilder = new DenseVectorFieldMapper.Builder(
+                INFERENCE_CHUNKS_RESULTS,
+                indexVersionCreated
+            );
+            SimilarityMeasure similarity = modelSettings.similarity();
+            if (similarity != null) {
+                switch (similarity) {
+                    case COSINE -> denseVectorMapperBuilder.similarity(DenseVectorFieldMapper.VectorSimilarity.COSINE);
+                    case DOT_PRODUCT -> denseVectorMapperBuilder.similarity(DenseVectorFieldMapper.VectorSimilarity.DOT_PRODUCT);
+                    default -> throw new IllegalArgumentException(
+                        "Unknown similarity measure for field [" + fieldName + "] in model settings: " + similarity
+                    );
+                }
+            }
+            Integer dimensions = modelSettings.dimensions();
+            if (dimensions == null) {
```

Comment:
Couldn't dims be calculated automatically if they weren't provided?

Reply:
They could! But it would signal that we're not getting what we expect from the inference results, which will always include them for dense vectors.